### ISSUE TITLE
RabbitMQ messages not processed

### ISSUE THREAD

**Issue by bovy89** [sensu/sensu#1385]
Hi, I'm experiencing an annoying problem with Sensu. Few minutes after sensu-server restart, messages in RabbitMQ are not more processed, and queue is constantly growing up.

There are no errors in Sensu and RabbitMQ.

Server Load  is very low and  a server with the same characteristics is handling four times the data of this one.

I've tried every version from 0.23.3 to 0.25.5.

OS: Centos 6.7

![img1](https://cloud.githubusercontent.com/assets/697408/17019579/4a85c7d0-4f3d-11e6-97c9-20477a064683.png)
![img2](https://cloud.githubusercontent.com/assets/697408/17019580/4a8a4f08-4f3d-11e6-965e-1f51284e8fc9.png)

**Comment by portertech** [comment]
@bovy89 what version of RabbitMQ and Erlang are you using?

**Comment by cwjohnston** [comment]
possibly related to sensu/sensu-transport#38

**Comment by portertech** [comment]
Unacked looks to be stuck at 2, which suggest that the Sensu consumer is hung and not acknowledging the messages its holding onto.

**Comment by bovy89** [comment]
@portertech 

erlang-18.1-1.el6.x86_64
rabbitmq-server-3.5.6-1.noarch

We have 6 virtual hosts on RabbitMQ. On each vhost there is a separated Sensu context (we use a central RabbitMQ setup to manage several Sensu).

Every Sensu server/client/api is using the same "core" components at the same version, but only this one is facing this issue.

**Comment by khalaan** [comment]
@bovy89 are you handling keepalives events? Is this just the keepalives queue or the results queue too?

**Comment by bovy89** [comment]
Both queue. I'm handling keepalives events.

**Comment by khalaan** [comment]
Mine always starts with keepalives falling behind first, then results follows about 10 minutes later. With only 1200 or so nodes on this instance.

**Comment by xyntrix** [comment]
We ran into this same issue with 8000 clients and 10  sensu-servers (overkill..).  We ended up setting the prefetch to 400 as a workaround.  While we'd notice 10 unacked messages on the keepalive or result queus, we would also notice the haproxy instance fronting RabbitMQ was consuming it's core frequently. As a response, we configured haproxy to make use of more cores on the box.  There seems/seemed to be a correlation between Sensu Server unable to recognize that message-processing degradation had occurred, or took so long to recognize it.. that we eventually would hit keepalive assertions -- I think some detection of this within the transport module is worth looking at to prevent Sensu from just "getting stuck" and therefore causing a cascade failure via issuing false keepalives.
